python/gstaichi/lang/ast/ast_transformer_utils.py

```diff
@@ -35,7 +43,10 @@ def __call__(self, ctx: "ASTTransformerContext", node: ast.AST):
                 raise GsTaichiSyntaxError(error_msg)
             info = ctx.get_pos_info(node) if isinstance(node, (ast.stmt, ast.expr)) else ""
             with impl.get_runtime().src_info_guard(info):
-                return method(ctx, node)
+                res = method(ctx, node)
+                if not hasattr(node, "ptr_source"):
```

Comment:
Having to rely on `hasattr` is usually a code smell.

Reply:
I see, but `ptr_source` is something added systematically added "manually" by gstaichi? Or it may or may not exist depending on some external conditional that is hard to foresee?

Reply:
Ok know I get it, `ast.parse` does not return a single node but a hierarchy I guess... so yes you would need to walk through the entire tree...

Reply:
Yeah, messes with typing for sure. I'm fairly OCD about typing, so I might poke at this at some point.